File: 2023/1/calibration_2.py
import re

def number_word_to_int(word):
  try:
    return int(word)
  except ValueError:
    return ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine'].index(word) + 1

def line_to_number(line):
  number_words = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
  all_matches = []
  for word in number_words:
    all_matches.extend(list(re.finditer(word, line)))
  sorted_matches = sorted(all_matches, key=lambda match: match.start())
  return number_word_to_int(sorted_matches[0].group()) * 10 + number_word_to_int(sorted_matches[-1].group())
  
f = open('input.txt')
lines = f.readlines()
sum = 0
for line in lines:
  number = line_to_number(line)
  sum += number

print(sum)

File: 2023/1/calibration_2.scala
import scala.io.Source
import scala.util.matching.Regex

def numberWordToInt(numberWord: String): Int = 
  try 
    numberWord.toInt
  catch 
    case e: NumberFormatException =>
      Array("one", "two", "three", "four", "five", "six", "seven", "eight", "nine").indexOf(numberWord) + 1

def lineToNumber(line: String): Int =
  val numberWords = Array("1", "2", "3", "4", "5", "6", "7", "8", "9", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine")
  val allMatches = numberWords.foldLeft(List[Regex.Match]()) { (acc, numberWord) =>
    acc ::: numberWord.r.findAllMatchIn(line).toList
  }
  val sortedMatches = allMatches.sortBy(m => m.start)
  val ints = sortedMatches.map(m => numberWordToInt(m.matched))
  ints.head * 10 + ints.last

@main def calibration() = 
  val bufferedSource = Source.fromFile("input.txt")
  val sum = bufferedSource.getLines.foldLeft(0) { (acc, line) =>
    acc + lineToNumber(line)
  }
  println(sum)

  bufferedSource.close